src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -0,0 +1,847 @@
+
+      // We abused the `out_buffer` to hold the result of the padding. Hence, we
+      // now have to copy the encrypted data over the padded plaintext data.
+      *out_size = encrypted.size();
+      Botan::copy_mem(std::span{out_buffer, *out_size}, encrypted);
+
+      return TSS2_RC_SUCCESS;
+   });
+   #else
+   BOTAN_UNUSED(pub_tpm_key, in_size, in_buffer, max_out_size, out_buffer, label, userdata);
+   return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+   #endif
+}
+
+/** Computation of an ephemeral ECC key and shared secret Z.
+ *
+ * According to the description in TPM spec part 1 C 6.1 a shared secret
+ * between application and TPM is computed (ECDH). An ephemeral ECC key and a
+ * TPM key are used for the ECDH key exchange.
+ * @param[in] key The key to be used for ECDH key exchange.
+ * @param[in] max_out_size the max size for the output of the public key of the
+ *            computed ephemeral key.
+ * @param[out] Z The computed shared secret.
+ * @param[out] Q The public part of the ephemeral key in TPM format.
+ * @param[out] out_buffer The public part of the ephemeral key will be marshaled
+ *             to this buffer.
+ * @param[out] out_size The size of the marshaled output.
+ * @param[in,out] userdata information.
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval USER_DEFINED user defined errors on failure.
+ */
+static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
+                              size_t max_out_size,
+                              TPM2B_ECC_PARAMETER* Z,
+                              TPMS_ECC_POINT* Q,
+                              BYTE* out_buffer,
+                              size_t* out_size,
+                              void* userdata) {
+   *out_size = 0;
+   #if defined(BOTAN_HAS_ECDH)
+   return thunk([&] {
+      BOTAN_ASSERT_NONNULL(key);
+
+      auto ccs = get(userdata);
+      if(!ccs) {
+         return TSS2_ESYS_RC_BAD_REFERENCE;
+      }
+
+      Botan::RandomNumberGenerator& rng = *ccs->get().rng;
+
+      // 1: Get TPM public key
+      const auto tpm_pub_inputs = Botan::TPM2::ecc_pubkey_from_tss2_public(key);
+      const auto tpm_sw_pubkey = Botan::ECDH_PublicKey(tpm_pub_inputs.first, tpm_pub_inputs.second);
+
+      const auto curve_order_byte_size = tpm_sw_pubkey.domain().get_p_bytes();
+
+      // 2: Generate ephemeral key
+      const auto eph_key = Botan::ECDH_PrivateKey(rng, tpm_sw_pubkey.domain());
+
+      // Get public key coordinates
+      const Botan::PointGFp& eph_pub_point = eph_key.public_point();
+      const auto& bn_x = eph_pub_point.get_affine_x();
+      const auto& bn_y = eph_pub_point.get_affine_y();
+
+      // Serialize affine coordinates into TPM2B_ECC_PARAMETER with Big Endian encoding.
+      // This ensures bn_{x,y}.bytes() <= curve_order_byte_size.
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->x, bn_x.serialize(curve_order_byte_size));
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->y, bn_y.serialize(curve_order_byte_size));
```

Comment:
If we extended the existing `as_span()` in `tpm2_utils.h` with a new overload that takes a `length` value for initialization, like so:

```C++
/// Set the size of @p data to @p lenght and construct a std::span
/// as a view into @p data
constexpr auto as_span(tpm2_buffer auto& data, size_t length) {
   BOTAN_ASSERT_NOMSG(length <= sizeof(data.buffer));
   data.size = static_cast<decltype(data.size)>(length);
   return as_span(data);
}
```

this could become much simpler (and omit temporary allocations -- for `BigInt::serialize()`):

```suggestion
      // Serialize public key coordinates
      const auto& eph_pub_point = eph_key.public_point();
      eph_pub_point.get_affine_x().serialize_to(Botan::TPM2::as_span(Q->x, curve_order_byte_size));
      eph_pub_point.get_affine_y().serialize_to(Botan::TPM2::as_span(Q->y, curve_order_byte_size));
```

Side note: the first overload of `copy_into()` in `tpm2_utils.h` could also make use of the new `as_span()` feature:

```C++
/// Copy the @p data into the TPM2 buffer @p dest, assuming that the
/// provided @p data is not larger than the capacity of the buffer.
template <tpm2_buffer T>
constexpr void copy_into(T& dest, std::span<const uint8_t> data) {
   copy_mem(as_span(dest, data.size()), data);
}
```